enhancements/network/egress-ip-per-destinations.md

```diff
@@ -0,0 +1,327 @@
+---
+title: egress-ip-per-destinations
+authors:
+- "@martinkennelly"
+reviewers:
+- "@trozet"
+- "@kyrtapz"
+approvers:
+- "@trozet"
+api-approvers:
+- "@joelspeed"
+creation-date: 2024-03-12
+last-updated: 2024-03-12
+tracking-link:
+- "https://issues.redhat.com/browse/SDN-4454"
+see-also:
+- "NA"
+replaces:
+- "NA"
+superseded-by:
+- "NA"
+---
+
+# Egress IP per destination
+## Summary
+Today, we can use an `EgressIP` to describe the source IP for a pod if it selected by an `EgressIP` custom resource (CR) selectors. This includes namespace
+and pod selectors. If multiple `EgressIP` CRs select the same set of pods, the behavior is undefined. This is because
+we cannot reliably choose which source IP to use.
+
+This enhancement proposes adding a new selector for when we want to apply the `EgressIP`. This new selector will only apply
+the `EgressIP` as the source IP to a set of pods communicating with destination IP if that destination IP is within predefined network CIDRs.
+
+If the new destination traffic selector is specified for all `EgressIP` CRs that selected a set of pods and the destination networks selected do not overlap, then we
+can allow a set of pods to have multiple source IPs depending on the destination address.
+
+## Motivation
+### User Stories
+
+As a cluster tenant, I want my application traffic to egress a particular interface depending on the traffic destination address and must
+have a distinct but consistent source IP for each interface.
+
+As a cluster administrator, I want to easily configure which destination networks should have an `EgressIP` because the destination
+networks may change, and I do not wish to have to update individual `EgressIP` custom resources for every update.
+
+### Goals
+- Allow a set of pods to have a consistent source IP depending on the destination
+- Support shared and local gateway (aka `routingViaHost=true`) modes in-order to reduce asymmetry between the modes
+- Backward compatibility with existing `EgressIP` API
+- If the new API isn't defined, preserve existing behaviour
+
+### Non-Goals
+- Change existing `EgressIP` behavior when the new field is not defined
+
+## Proposal
+
+A new `TrafficSelector` field is added to `EgressIP` spec which uses Kubernetes label selector semantics
+to select a new CRD called `EgressIPTraffic` which contains a list of destination networks. This list of networks defines when an `EgressIP` should
```

Comment:
I think it is implied, but to confirm, these aspects of EgressIP do not change as a result of this enhancement (either with or without a traffic selector applied):
* Traffic egressing under an EIP (either without traffic selector or as a result of being selected) will be load balanced across multiple IP addresses defined within that EIP.
* Placement of IP addresses from an EIP is based on egress-assignable nodes and interfaces with appropriate subnets to host the IP.

One aspect of this is that the user may design networking connectivity as desired/needed. The multiple egress networks (EIP) do not necessarily need to be connected to the same node(s).

Reply:
>Traffic egressing under an EIP (either without traffic selector or as a result of being selected) will be load balanced across multiple IP addresses defined within that EIP.

yes

> Placement of IP addresses from an EIP is based on egress-assignable nodes and interfaces with appropriate subnets to host the IP.

yes